add docs for first run experiments, update document titles (#372)
mozilla/experimenter-docs: docs/deep-dives/mobile/first-run-experiments.mdx

---
id: mobile-first-run-experiments
title: First Run Experiments
slug: /mobile-first-run-experiments
---

> First run experiments are incredibly important to the experimentation ecosystem, but their functionality comes with some key differences from standard experiments. It is _highly_ advised that experiment owners read through this documentation before writing first run experiments.

Reviewer: I would *strongly* suggest that you spell `First Run Experiments` with consistent capitalization, as so we can refer to an experiment as a "First Run Experiment", and everyone knows what it is.

:::note

> The concept of first run experiments is currently only supported on mobile applications, such as Firefox for Android/iOS and Focus.

Reviewer 2: Could we link to docs on our capabilities for Desktop? Because we do support something functionally equivalent there

Author: Oh where is our desktop documentation?

Reviewer 2: I assumed we had some when I wrote that comment... But now all I can find is [this](https://experimenter.info/desktop-feature-api/#can-i-use-this) which claims we support it for desktop

Author: Oh so maybe I just take this whole note out then

Author: Or reword it to explain that this particular set of docs is only pertaining to our mobile applications

:::

## What is a first run experiment?

A first run experiment is one that needs to enroll clients on their first launch of the application. All first run experiments bundled with a particular app version would be evaluated for enrollment on the very first run of the application after it is installed.

An experiment can be marked as a first run experiment by checking the "First Run Experiment" checkbox while editing the experiment in the audience page.

> <img title="First run checkbox" alt="The first run experiment checkbox is shown on the audience page of an experiment being edited"

Reviewer: Alt text=" 😻 "

Author: I try my best to make sure accessibility is considered and covered where possible 😊

     src="/img/deep-dives/first-run-experiment-checkbox.png" height="500" style={{marginBottom: '10px'}}/>

Similarly to sticky enrollment, certain advanced targeting configurations can require an experiment to be first run.

<img title="First run checkbox required" alt="The first run experiment checkbox is selected and cannot be edited because the advanced targeting configuration selected requires the experiment to be a first run experiment."
     src="/img/deep-dives/first-run-required.png" height="500" style={{marginBottom: '10px'}}/>

The differences between first run experiments and standard experiments show themselves in a few key areas, described below.

### Application Bundling

Unlike traditional experiments, first run experiments must be bundled with the app. This allows Nimbus to start quickly without needing to make a call to remote settings to fetch the live experiments.

> This means that each change to which first-run experiments are live must be met with a code change to add or remove experiments from the bundled file. This would be a very frustrating process to keep up with manually, so we've created an [update-experiments github action](https://github.com/mozilla-mobile/update-experiments) to automate it. Below is an example job configuration for this action.

Reviewer: ```suggestion
This means that any new First Run Experiments are added to the application's source code repository. This happens automatically via the [`update-experiments` Github action](https://github.com/mozilla-mobile/update-experiments). Below is an example job configuration for this action.
```

          token: ${{ secrets.GITHUB_TOKEN }}
          path: $DIRECTORY
          branch: automation/update-nimbus-experiments
          commit-message: "update initial_experiments.json based on the current first-run experiments in experimenter"
          title: "Update initial experiments JSON for Nimbus"
          body: "This (automated) PR updates the initial_experiments.json on the `main` branch"
          delete-branch: true
```

The only file type currently supported is JSON, and this is usually named `initial_experiments.json`. Some examples of how this workflow and file are used can be seen in [this PR](https://github.com/mozilla-mobile/firefox-ios/pull/12872/files) and [the `initial_experiments.json` file](https://github.com/mozilla-mobile/firefox-ios/blob/main/Client/Experiments/initial_experiments.json) in the Firefox iOS repository.

### Enrollment

For traditional experiments, enrollment is a short period at the beginning of a live experiment's lifecycle where the experiment can enroll new users. The Nimbus SDK does not do any date calculations when determining whether an experiment is enrolling users, instead relying on the `isEnrollmentPaused` experiment field.

> Because first run experiments are bundled with the application, that field will remain `true` in the application bundle for the duration of the release. As a result, **even after enrollment is paused within experimenter, first run users will still be evaluated and enrolled.**

Reviewer: ```suggestion
Because first run experiments are bundled with the application, that field will remain `true` in the application bundle for the duration of the release. As a result, **even after enrollment is paused remotely via `experimenter`, First Run Experiments on the app's first run will still be evaluated and enrolled.**
```

> Because first run experiments are bundled with the application, that field will remain `true` in the application bundle for the duration of the release. As a result, **even after enrollment is paused within experimenter, first run users will still be evaluated and enrolled.**

Author: Yep that's a good callout, I'll add that

When defining the enrollment length and experiment lifetime, it is wise to take this fact into account in planning. As an example, if a first run experiment should run for four weeks, its end date should be four weeks after the date of release for the version it is targeting.

### Advanced Targeting

> Many use cases for first run experiments require advanced targeting configurations. To add a new advanced targeting configuration, a pull request should be opened to the Experimenter repository, adding the configuration to [this](https://github.com/mozilla/experimenter/blob/main/app/experimenter/targeting/constants.py) file.

Reviewer: ```suggestion
Many use cases for first run experiments require advanced targeting configurations. To add a new advanced targeting configuration, a pull request should be opened to the Experimenter repository, adding the configuration to [this file](https://github.com/mozilla/experimenter/blob/main/app/experimenter/targeting/constants.py).
```
